subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformer.java

```diff
@@ -141,33 +142,30 @@ public int hashCode() {
     }
 
     private static class TransformServiceLookup implements ServiceLookup {
-        private final File inputFile;
-        private final File outputDir;
-        private final Object parameters;
-        private final ArtifactTransformDependencies artifactTransformDependencies;
+        private final ImmutableList<InjectionPoint> injectionPoints;
 
         public TransformServiceLookup(File inputFile, File outputDir, @Nullable Object parameters, @Nullable ArtifactTransformDependencies artifactTransformDependencies) {
-            this.inputFile = inputFile;
-            this.outputDir = outputDir;
-            this.parameters = parameters;
-            this.artifactTransformDependencies = artifactTransformDependencies;
+            ImmutableList.Builder<InjectionPoint> builder = ImmutableList.builder();
+            builder
+                .add(new InjectionPoint(Workspace.class, File.class, outputDir))
+                .add(new InjectionPoint(PrimaryInput.class, File.class, inputFile));
```

Comment:
If we had that meta-annotation, we could get the potential target types directly from there.